src/citrine/resources/file_link.py

```diff
@@ -547,10 +575,77 @@ def delete(self, file_link: FileLink):
             Resource referencing the external file.
 
         """
+        if self._is_external_url(file_link.url):
+            raise ValueError(f"Only local resources can be deleted; passed URL {file_link.url}")
+        if not self._validate_filelink_url(file_link.url):
+            raise ValueError(f"URL was malformed for local resources; passed URL {file_link.url}")
         split_url = file_link.url.split('/')
-        assert split_url[-2] == 'versions' and split_url[-4] == 'files', \
-            "File URL is expected to end with '/files/{{file_id}}/version/{{version id}}', " \
-            "but FileLink instead has url {}".format(file_link.url)
         file_id = split_url[-3]
         data = self.session.delete_resource(self._get_path(file_id))
         return Response(body=data)
+
+    def _resolve_file_link(self, identifier: Union[str, UUID]) -> FileLink:
+        """Generate the FileLink object referenced by the passed argument."""
+        if isinstance(identifier, FileLink):
+            return identifier
+        try:  # Check if the string is actually a UUID
+            if isinstance(identifier, str):
+                identifier = UUID(identifier)
+        except ValueError:
+            pass
+        if isinstance(identifier, UUID):
+            # Assume it's the file UUID on platform
+            hit = next((f for f in self.list() if str(identifier) in f.url), None)
+            if hit is None:
+                raise ValueError(f"Found no file associated with UUID {identifier}")
+            else:
+                return hit
+        elif isinstance(identifier, str) and self._is_external_url(identifier):
+            # Assume it's an absolute URL
+            filename = urlparse(identifier).path.split('/')[-1]
+            file_dict = {
+                'url': identifier,
+                'filename': filename,
+                'type': GEMDFileLink.typ
+            }
+            return FileLink.build(file_dict)
+        elif isinstance(identifier, str) and self._validate_filelink_url(identifier):
+            # It's an on-platform URL
+            hit = next((f for f in self.list() if identifier == f.url), None)
```

Comment:
Why do you have to iterate with .list here? This looks inefficient.

Reply:
There's no direct get-by-url, but doing the parse here makes sense.  I think I was trying to reduce the amount of URL twiddling that was everywhere in this module.  Will update.

Reply:
Reworked to use get, though annoyingly get was broken.  The files routes have different call structure than other calls in the API.

In particular see [files.get](https://docs.corp.citrine.io/shire/pippin/api/index.html#tag/Files/paths/~1projects~1{projectId}~1datasets~1{datasetId}~1files~1{fileId}/get) and [files.versions.get](https://docs.corp.citrine.io/shire/pippin/api/index.html#tag/Files/paths/~1projects~1{projectId}~1datasets~1{datasetId}~1files~1{fileId}~1versions~1{versionId}/get)